src/webgpu/api/validation/render_pipeline/inter_stage.spec.ts

```diff
@@ -1,10 +1,337 @@
+  });
+
+g.test('interpolation_sampling')
+  .desc(
+    `Tests that validation should fail when interpolation sampling of vertex output and fragment input at the same location doesn't match.`
+  )
+  .params(u =>
+    u.combine('isAsync', [false, true]).combineWithParams([
+      // default is @interpolate(perspective, center)
+      { output: '@interpolate(perspective) ', input: '@interpolate(perspective) ' },
+      {
+        output: '@interpolate(perspective) ',
+        input: '@interpolate(perspective, center) ',
+        _success: true,
+      },
+      { output: '@interpolate(linear, center) ', input: '@interpolate(linear) ', _success: true },
+      { output: '@interpolate(flat) ', input: '@interpolate(flat) ' },
+      { output: '@interpolate(perspective) ', input: '@interpolate(perspective, sample) ' },
+      { output: '@interpolate(perspective, center) ', input: '@interpolate(perspective, sample) ' },
+      {
+        output: '@interpolate(perspective, center) ',
+        input: '@interpolate(perspective, centroid) ',
+      },
+      { output: '@interpolate(perspective, centroid) ', input: '@interpolate(perspective) ' },
+    ])
+  )
+  .fn(async t => {
+    const { isAsync, output, input, _success } = t.params;
+
+    const descriptor = t.getDescriptorWithStates(
+      t.getVertexStateWithOutputs([`@location(0) ${output}vout0: f32`]),
+      t.getFragmentStateWithInputs([`@location(0) ${input}fin0: f32`])
+    );
+
+    t.doCreateRenderPipelineTest(isAsync, _success ?? output === input, descriptor);
+  });
+
+g.test('max_shader_variable_location')
+  .desc(
+    `Tests that validation should fail when there is location of user-defined output/input variable >= device.limits.maxInterStageShaderVariables`
+  )
+  .params(u =>
+    u
+      .combine('isAsync', [false, true])
+      // user defined variable location = maxInterStageShaderVariables + locationDelta
+      .combine('locationDelta', [0, -1])
+  )
+  .fn(async t => {
+    const { isAsync, locationDelta } = t.params;
+    const location = t.device.limits.maxInterStageShaderVariables + locationDelta;
+
+    const descriptor = t.getDescriptorWithStates(
+      t.getVertexStateWithOutputs([`@location(${location}) vout0: f32`]),
+      t.getFragmentStateWithInputs([`@location(${location}) fin0: f32`])
+    );
+
+    t.doCreateRenderPipelineTest(
+      isAsync,
+      location < t.device.limits.maxInterStageShaderVariables,
+      descriptor
+    );
+  });
+
+g.test('max_components_count,output')
+  .desc(
+    `Tests that validation should fail when scalar components of all user-defined outputs > max vertex shader output components.`
+  )
+  .params(u =>
+    u.combine('isAsync', [false, true]).combineWithParams([
+      // Number of user-defined output scalar components in test shader = device.limits.maxInterStageShaderComponents + numScalarDelta - 4. (builtin position output counts for space as well)
```

Comment:
With the spec we settled on, you shouldn't need the `- 4`. The value of `maxInterStageShaderComponents` itself is supposed to reserve room for `@builtin(position)`. (Dawn doesn't implement that yet.)